### ISSUE TITLE
1.0 Release Features

### ISSUE THREAD

**Issue by timbod7** [adl-lang/adl#101]
ADL has been actively used in a range of projects for more than 4 years. It's worked well, though usage has suggested some improvements that are difficult to make in a backwards compatible fashion. Hence it is proposed that these changes be made on the `master` branch leading to a 1.0 release. In the meantime, bug fixes to exist releases can be made on a `release-0.x` branch.

Proposed 1.0 features include:

- [x] rename `Void` primitive to `Unit`
- [x] Address issues with Unit type in java (https://github.com/timbod7/adl/issues/56)
- [x] Change `sys.types.Map` serialization to a list of map entries (not pairs)
- [ ] Fix java builders to avoid  runtime errors on incomplete construction
- [x] Make enums in typescript be type level strings

Feel free to propose other changes in the thread.

**Comment by timbod7** [comment]
Remove `Error<T>` type from sys.types (it's been replaced with `Result<T,E>`.

**Comment by gmhta** [comment]
- [ ] packaging system.

**Comment by timbod7** [comment]
> packaging system

Interesting! Need's some thought and discussion...

**Comment by gmhta** [comment]
- [ ] lsp

**Comment by PaulThompson** [comment]
More options for union serialisation.
I particular ability to have a unit type field of a union serialise as users choice of `null`, `{}`, `[]` (or the existing: field label)
So that the field can change in future to become any `Nullable<T>`, `struct` or `Vector<T>`

**Comment by PaulThompson** [comment]
Ability to opt out of unions code-generating as an enum.
It complicates generic tools because unions that generate as enum have to be treated differently and during development a type can flip between being a normal union and being an enum when a field is added/removed.

**Comment by PaulThompson** [comment]
Implement Nullable<T> for c++ (target c++17 and use std::optional)

**Comment by bitc** [comment]
I have added a comment on #97 (regarding strict fields). I bring this up here, because this is something that could affect backwards compatibility.

**Comment by PaulThompson** [comment]
In c++ - given ADL module 
`module aa.bb.cc {};`
it should be changed (non-backwards compatible) to generate c++ code in namespaces:

```
namespace aa {
namespace bb {
namespace cc {
// ...
}}}
```

Currently it generates to namespace:
```
namespace ADL {
namespace aa {
namespace bb {
namespace cc {
// ...
}}}
```
which IIUC isn't requested or controlled by the caller.

**Comment by timbod7** [comment]
> it should be changed (non-backwards compatible)
 
I don't think this needs to be a backward imcompatible change - don't we just need a flag for c++ that controls into which namespace ADL code is generated? We have equivalent flags for most other target languages.

**Comment by timbod7** [comment]
Drop the baked in json support in the C++ backend, and use nlohmann json.

**Comment by timbod7** [comment]
There's now a milestone for "release 1.0", which will be added to issues as they make the cur.

**Comment by timbod7** [comment]
@gmhta I'm leaving:

* lsp
* packaging system

out of the 1.0 release milestone,  as I think they can be added when time permits without breaking existing usage.

**Comment by timbod7** [comment]
@paul-thompson-helix 

nullable support in c++ has already landed in f42788ea4d2be23fe100dbbd9da5e8a6204a2df2

**Comment by timbod7** [comment]
Also see the release-1.0 branch for progress on these issues.
